odxtools/minmaxlengthtype.py

```diff
@@ -34,113 +34,136 @@ def __post_init__(self):
-        byte_length = self._minimal_byte_length_of(internal_value)
-
-        # The coded value must have at least length min_length
-        if byte_length < self.min_length:
-            raise EncodeError(f"The internal value {internal_value} is only {byte_length} bytes"
-                              f" long but the min length is {self.min_length}")
-        # The coded value must not have a length greater than max_length
-        if self.max_length and byte_length > self.max_length:
-            raise EncodeError(f"The internal value {internal_value} requires {byte_length}"
-                              f" bytes, but the max length is {self.max_length}")
-
-        value_byte = self._to_bytes(
-            internal_value,
-            bit_position=0,
-            bit_length=8 * byte_length,
-            base_data_type=self.base_data_type,
-            is_highlow_byte_order=self.is_highlow_byte_order,
-        )
-
-        if encode_state.is_end_of_pdu or byte_length == self.max_length:
-            # All termination types may be ended by the PDU
-            return value_byte
+        bit_length = 8 * self.max_length if self.max_length is not None else 8 * len(internal_value)
+        bit_length = min(8 * len(internal_value), bit_length)
+        value_bytes = bytearray(
+            self._to_bytes(
+                internal_value,
+                bit_position=0,
+                bit_length=bit_length,
+                base_data_type=self.base_data_type,
+                is_highlow_byte_order=self.is_highlow_byte_order,
+            ))
+
+        # TODO: ensure that the termination delimiter is not
+        # encountered within the encoded value.
+
+        if encode_state.is_end_of_pdu or len(value_bytes) == self.max_length:
+            # All termination types may be ended by the end of the PDU
+            # or once reaching the maximum length. In this case, we
+            # must not add the termination sequence
+            pass
         else:
-            termination_char = self.__termination_character()
-            if self.termination == "END-OF-PDU":
-                termination_char = b''
-            odxassert(
-                termination_char is not None,
-                f"MinMaxLengthType with termination {self.termination}"
-                f"(min: {self.min_length}, max: {self.max_length}) failed encoding {internal_value}"
-            )
-            return value_byte + termination_char
+            termination_sequence = self.__termination_sequence()
+
+            # ensure that we don't try to encode an odd-length
+            # value when using a two-byte terminator
+            odxassert(len(value_bytes) % len(termination_sequence) == 0)
+
+            value_bytes.extend(termination_sequence)
+
+        if len(value_bytes) < self.min_length:
+            raise EncodeError(f"Encoded value for MinMaxLengthType "
+                              f"must be at least {self.min_length} bytes long. "
```

Comment:
* Suggestion: use odxassert instead.
* min_length => is this in bytes or in characters?

Reply:
> Suggestion: use odxassert instead.

I think that at the other occasions where we catch an illegal parameter value, we raise an EncodeError as well?

> min_length => is this in bytes or in characters?

the spec says bytes. (Note that I think that this is a bad idea for anything other than ASCII or UTF-8 strings. IMO, they should have restricted this kind of DOP to strings, removed the non-null termination options and called it something like "CStringDOP"...)

> No f"(Is: {len(value_bytes)} bytes.)" ?

I guess that could be useful to find a value which works (how to determine the the length of an encoded from a physical value is not obvious for me...)